Framework/PythonInterface/mantid/FunctionWrapper.py

```diff
@@ -0,0 +1,480 @@
+           its list of functions, so having a pure list of functions.
+           This makes it possible to index and iterate all the functions
+           and use tieAll() and untieAll().
+           Not to be used with a mixture of product and sum 
+           composite functions, because the arithmetic 
+           may no longer be correct.
+           The return value is not independent of self.
+       """
+       # If there are no composite functions, do nothing
+       needToFlatten = False
+       for i in range(0, len(self)):
+          if not needToFlatten and isinstance(self[i],CompositeFunctionWrapper):
+             needToFlatten = True
+             
+       if not needToFlatten :
+          return self
+       
+       # Now we know there is a composite function.
+       if isinstance(self,ProductFunctionWrapper):
+          flatSelf = ProductFunctionWrapper()
+       else:
+          flatSelf = CompositeFunctionWrapper()
+          
+       for i in range(0, len(self)):
+          if isinstance(self[i],CompositeFunctionWrapper):
+             currentFunction = self[i].flatten()
+             for j in range(0, len(currentFunction)):
+                flatSelf.fun.add(currentFunction[j].fun)
+          else:
+             flatSelf.fun.add(self[i].fun)
+             
+       return flatSelf          
+ 
+class ProductFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       self.pureAddition = False
+       self.pureMultiplication = True
+       return self.initByName("ProductFunction", *args)
+       
+class ConvolutionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Convolution Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       self.pureAddition = False
+       self.pureMultiplication = False
+       return self.initByName("Convolution", *args)
+       
+class MultiDomainFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args, **kwargs):
+       """ Called when creating an instance
+           It should not be called directly
+       """
+       # Assume it's not safe to flatten
+       self.pureAddition = False
+       self.pureMultiplication = False
+       
+       # Create and populate with copied functions
+       self.initByName("MultiDomainFunction", *args)
+               
+       # Tie the global parameters
+       if 'Global' in kwargs:
+          list = kwargs['Global']
+          for name in list:
+             self.tieAll(name)
+       
+       # Set domain indices: 1 to 1       
+       for i in range(0, len(self)):
+          self.fun.setDomainIndex(i, i)
+          
+    def nDomains (self):
```

Comment:
I see. 

I'm finding http://www.python-course.eu/python3_properties.php a good introduction to properties.